### PR TITLE
drivers: tzc380: add tzc380 driver

### PR DESCRIPTION
**Pull request by MrVan**
Add tzc380 driver support.

The usage:
Use tzc_init(vaddr_t base) to get the tzc380 configuration.
Use tzc_configure_region to configure the memory region,
such as "tzc_configure_region(5, 0x4e000000,
TZC_ATTR_REGION_SIZE(TZC_REGION_SIZE_32M) | TZC_ATTR_REGION_EN_MASK |
TZC_ATTR_SP_S_RW);"

An example for 6ULEVK:
```
        tzc_init(addr);

        tzc_configure_region(0, 0x00000000, TZC_ATTR_REGION_SIZE(TZC_REGION_SIZE_4G) | TZC_ATTR_REGION_EN_MASK | TZC_ATTR_SP_ALL);
        tzc_configure_region(1, 0x80000000, TZC_ATTR_REGION_SIZE(TZC_REGION_SIZE_2G) | TZC_ATTR_REGION_EN_MASK | TZC_ATTR_SP_NS_RW);
        tzc_configure_region(2, 0x9e000000, TZC_ATTR_REGION_SIZE(TZC_REGION_SIZE_32M) | TZC_ATTR_REGION_EN_MASK | TZC_ATTR_SP_S_RW);
        tzc_configure_region(3, 0x9e000000, TZC_ATTR_REGION_SIZE(TZC_REGION_SIZE_2M) | TZC_ATTR_REGION_EN_MASK | TZC_ATTR_SP_ALL);

        tzc_dump_state();
```

### CODE REVIEW

File: core/include/drivers/tzc380.h [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+#define TZC_REGION_SIZE_256P	(0x39)
+#define TZC_REGION_SIZE_512P	(0x3a)
+#define TZC_REGION_SIZE_1E	(0x3b)
+#define TZC_REGION_SIZE_2E	(0x3c)
+#define TZC_REGION_SIZE_4E	(0x3d)
+#define TZC_REGION_SIZE_8E	(0x3e)
+#define TZC_REGION_SIZE_16E	(0x3f)
+
+#define TZC_REGION_SIZE_SHIFT	(0x1)
+#define TZC_REGION_SIZE_MASK	(0x3f << 1)
```
Comment (etienne-lms) [inline core/include/drivers/tzc380.h]: `#define TZC_REGION_SIZE_MASK	GENMASK_32(7, 1);`

File: core/include/drivers/tzc380.h [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+#define TZC_ATTR_REGION_SIZE(s)	((s) << TZC_REGION_SIZE_SHIFT)
+
+#define TZC_ATTR()
```
Comment (etienne-lms) [inline core/include/drivers/tzc380.h]: empty macro ?

File: core/include/drivers/tzc380.h [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	TZC_ACTION_ERR = 1,
+	TZC_ACTION_INT = 2,
+	TZC_ACTION_ERR_INT = (TZC_ACTION_ERR | TZC_ACTION_INT)
+};
+
+
+#define TZC_SP_NS_W		(1 << 0)
+#define TZC_SP_NS_R		(1 << 1)
+#define TZC_SP_S_W		(1 << 2)
+#define TZC_SP_S_R		(1 << 3)
```
Comment (etienne-lms) [inline core/include/drivers/tzc380.h]: `#define TZC_SP_NS_W		BIT(0)` and so on

File: core/include/drivers/tzc380.h [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+#define ACTION_RV_SHIFT		0
+#define ACTION_RV_MASK		0x3
+#define  ACTION_RV_LOWOK	0x0
+#define  ACTION_RV_LOWERR	0x1
+#define  ACTION_RV_HIGHOK	0x2
+#define  ACTION_RV_HIGHERR	0x3
+
+/* Speculation is enabled by default. */
+#define SPECULATION_CTRL_WRITE_DISABLE	(1 << 1)
+#define SPECULATION_CTRL_READ_DISABLE	(1 << 0)
```
Comment (etienne-lms) [inline core/include/drivers/tzc380.h]: prefer use `BIT(1)` and `BIT(0)`.

File: core/include/drivers/tzc380.h [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __DRIVERS_TZC380_H
+#define __DRIVERS_TZC380_H
+
+#include <stdint.h>
+#include <types_ext.h>
+#include <trace_levels.h>
```
Comment (etienne-lms) [inline core/include/drivers/tzc380.h]: shift type_ext.h below trace_levels.h

File: core/drivers/tzc380.c [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+			continue;
+
+		DMSG("\n");
+		DMSG("region %d", n);
+		temp_32reg = tzc_read_region_base_low(tzc.base, n);
+		temp_32reg_h = tzc_read_region_base_high(tzc.base, n);
+		DMSG("region_base: 0x%08x%08x", temp_32reg_h, temp_32reg);
+		temp_32reg = tzc_read_region_attributes(tzc.base, n);
+		DMSG("region sp: %x", temp_32reg >> TZC_ATTR_SP_SHIFT);
+		DMSG("region size: %x\n", (temp_32reg & TZC_REGION_SIZE_MASK) >> TZC_REGION_SIZE_SHIFT);
```
Comment (etienne-lms) [inline core/drivers/tzc380.c]: style: exeeds 80 chars

File: core/drivers/tzc380.c [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (etienne-lms) [inline core/drivers/tzc380.c]: tzc400.c also includes an ARM-TF notice. Should it be dumped here also ?
` * Copyright (c) 2014, ARM Limited and Contributors. All rights reserved.`

File: core/include/drivers/tzc380.h (hunk 5 above)
Comment (MrVan) [inline core/include/drivers/tzc380.h]: Fix in next version.

File: core/include/drivers/tzc380.h (hunk 4 above)
Comment (MrVan) [inline core/include/drivers/tzc380.h]: Fix to use BIT for all the bits things in next version

File: core/include/drivers/tzc380.h (hunk 1 above)
Comment (MrVan) [inline core/include/drivers/tzc380.h]: Fix in next version.

File: core/include/drivers/tzc380.h (hunk 2 above)
Comment (MrVan) [inline core/include/drivers/tzc380.h]: Drop this in next version. Thanks.

File: core/drivers/tzc380.c (hunk 7 above)
Comment (MrVan) [inline core/drivers/tzc380.c]: There is no tzc380 support in ARM-TF.

File: core/drivers/tzc380.c (hunk 6 above)
Comment (MrVan) [inline core/drivers/tzc380.c]: Fix in next version.

File: core/drivers/tzc380.c (hunk 7 above)
Comment (etienne-lms) [inline core/drivers/tzc380.c]: I was thinking of the tzc400.c from optee which is almost the same as this proposed tzc380.c.

Comment (MrVan) [inline core/drivers/tzc380.c]: I take tzc400.c as a reference to implement tzc380 support. Simliar functions are implemented in tzc380. But code is not copied from tzc400.c. I am not sure the copyright is needed.

Comment (etienne-lms) [inline core/drivers/tzc380.c]: ok. discard my comments. thanks.

**Comment by MrVan** [comment]
Updated with comments addressed.

File: core/drivers/tzc380.c [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+		DMSG("\n");
+		DMSG("region %d", n);
+		temp_32reg = tzc_read_region_base_low(tzc.base, n);
+		temp_32reg_h = tzc_read_region_base_high(tzc.base, n);
+		DMSG("region_base: 0x%08x%08x", temp_32reg_h, temp_32reg);
+		temp_32reg = tzc_read_region_attributes(tzc.base, n);
+		DMSG("region sp: %x", temp_32reg >> TZC_ATTR_SP_SHIFT);
+		DMSG("region size: %x\n", (temp_32reg & TZC_REGION_SIZE_MASK)
+		     >> TZC_REGION_SIZE_SHIFT);
```
Comment (etienne-lms) [inline core/drivers/tzc380.c]: minor style: end line175 with the operator `>>`.

**Comment by MrVan** [comment]
@etienne-lms Updated.

**Comment by etienne-lms** [comment]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by MrVan** [comment]
@etienne-lms tags already applied, thanks:)

**Comment by jforissier** [comment]
The failed build status is due to a timeout (Travis seems to be slower than usual today), the final check was almost complete so I'm merging this regardless.

**Pull request merged**